# Exemplo 1

In [ ]:
import numpy as np
from scipy.stats import chi2, poisson

# Geracao dos dados se basendo-se em Poisson
dados = np.random.poisson(lam=5, size=1000)

n = len(dados)
alpha = 0.05    #Nível de significancia

# Estimativa de lambda
lambda_estimado = np.mean(dados)

# Classes: 0,1,2,3,4,5,6,7, >=8
k = 9
frequencias = [0] * k

for v in dados:
    if v >= 8:
        frequencias[8] += 1
    else:
        frequencias[v] += 1
        
# Frequencias esperadas
E = []
for i in range(8):
    prob = poisson.pmf(i, lambda_estimado)
    E.append(n * prob)

# Classe >=8
prob_tail = 1 - sum(poisson.pmf(i, lambda_estimado) for i in range(8))
E.append(n * prob_tail)

# Qui-quadrado
chi2_calculado = sum((f - e)**2 / e for f, e in zip(frequencias, E))

# Graus de liberdade
gl = k - 2

# Valor critico
chi2_critico = chi2.ppf(1 - alpha, gl)

# Saidas
print("Lambda estimado:", lambda_estimado)
print("Frequencias observadas:", frequencias)
print("Frequencias esperadas:", [round(e,2) for e in E])
print("Qui-quadrado:", chi2_calculado)
print("Valor critico:", chi2_critico)

if chi2_calculado <= chi2_critico:
    print("Nao se rejeita H0 (Poisson)")
else:
    print("Rejeita-se H0")



Lambda estimado: 5.012
Frequencias observadas: [7, 36, 84, 132, 170, 190, 146, 105, 130]
Frequencias esperadas: [np.float64(6.66), np.float64(33.37), np.float64(83.62), np.float64(139.7), np.float64(175.04), np.float64(175.46), np.float64(146.57), np.float64(104.95), np.float64(134.63)]
Qui-quadrado: 2.1622548840158893
Valor critico: 14.067140449340169
Nao se rejeita H0 (Poisson)


Teste de Aderência para Distribuição de Poisson

Neste exemplo, o código gera 1000 números aleatórios seguindo uma distribuição de Poisson com $\lambda = 5$.

**O que o código faz:** Ele tenta "redescobrir" se esses dados pertencem a uma distribuição de Poisson. Ele calcula a média dos dados para estimar o $\lambda$ (Lambda), define 9 categorias (de 0 a 7, e uma categoria para valores $\ge 8$) e compara quantas vezes cada número apareceu (frequência observada) com o que seria esperado matematicamente.

**Resultado:** Como os dados foram gerados artificialmente como Poisson, o valor de Qui-quadrado calculado ($2.16$) é muito menor que o valor crítico ($14.07$).

**Conclusão:** Não se rejeita $H_0$, ou seja, os dados realmente seguem a distribuição de Poisson.

#   Exemplo 2

In [2]:
import math
from scipy.stats import chi2

f = [50, 60, 70, 80, 90, 110, 120, 140, 280]

n = sum(f)
k = len(f)
alpha = 0.05
lambda_estimado = 6

E = []

for i in range(8):
    prob = math.exp(-lambda_estimado) * lambda_estimado**i / math.factorial(i)
    E.append(n * prob)

# Classe >= 8
prob_tail = 1 - sum(E) / n
E.append(n * prob_tail)

chi2_calculado = sum((fo - fe)**2 / fe for fo, fe in zip(f, E))

gl = k - 2
chi2_critico = chi2.ppf(1 - alpha, gl)

print("Frequencias esperadas:", [round(e, 2) for e in E])
print("Qui-quadrado:", round(chi2_calculado, 2))
print("Valor critico:", round(chi2_critico, 2))

if chi2_calculado > chi2_critico:
    print("Rejeita-se H0")
else:
    print("Nao se rejeita H0")
    



Frequencias esperadas: [2.48, 14.87, 44.62, 89.24, 133.85, 160.62, 160.62, 137.68, 256.02]
Qui-quadrado: 1106.26
Valor critico: 14.07
Rejeita-se H0


Rejeição da Hipótese de Poisson

Este exemplo utiliza dados manuais inseridos em uma lista f.

**O que o código faz:** Ele define frequências observadas que não foram geradas por uma função de Poisson e tenta forçar um teste contra um $\lambda = 6$.

**Resultado:** As frequências esperadas calculadas pelo modelo são muito diferentes das observadas (ex: esperado 2.48 para a primeira categoria, mas observado 50). Isso gera um valor de Qui-quadrado astronômico ($1106.26$).

**Conclusão:** Rejeita-se $H_0$. Os dados fornecidos não seguem uma distribuição de Poisson com $\lambda = 6$.

# Exemplo 3

In [3]:
import numpy as np
from scipy.stats import chi2

# Geracao dos dados
np.random.seed(42)
dados = np.random.exponential(scale=2, size=1000)

n = len(dados)
alpha = 0.05

# Estimativa de lambda
media = np.mean(dados)
lambda_estimado = 1 / media

print("Media:", round(media, 3))
print("Lambda estimado:", round(lambda_estimado, 3))

# Intervalos equiprovaveis (aprox. 20%)
intervalos = [0, 0.45, 1.02, 1.83, 3.22, np.inf]

k = 5
frequencias = [0] * k

# Contagem das frequencias observadas
for v in dados:
    for i in range(k):
        if intervalos[i] <= v < intervalos[i+1]:
            frequencias[i] += 1
            break

print("Frequencias observadas:", frequencias)

# Frequencias esperadas (equiprovaveis)
E = [n / k] * k

# Calculo do qui-quadrado
chi2_calculado = sum((f - e)**2 / e for f, e in zip(frequencias, E))

# Graus de liberdade
gl = k - 2

# Valor critico
chi2_critico = chi2.ppf(1 - alpha, gl)

print("Qui-quadrado:", round(chi2_calculado, 2))
print("Valor critico:", round(chi2_critico, 2))

if chi2_calculado <= chi2_critico:
    print("Nao se rejeita H0 (distribuicao exponencial)")
else:
    print("Rejeita-se H0")



Media: 1.945
Lambda estimado: 0.514
Frequencias observadas: [225, 196, 192, 188, 199]
Qui-quadrado: 4.25
Valor critico: 7.81
Nao se rejeita H0 (distribuicao exponencial)


Teste de Aderência para Distribuição Exponencial

Aqui, o código lida com dados contínuos gerados por uma distribuição exponencial com escala 2 ($\lambda \approx 0.5$).

**O que o código faz:** Como a distribuição exponencial é contínua, o código divide os dados em 5 "baldes" (intervalos) baseados em probabilidades equiprováveis (cada intervalo deveria ter 20% dos dados).

**Cálculo:** Ele estima o parâmetro $\lambda$ através da média dos dados ($1 / \text{média}$).Resultado: O Qui-quadrado calculado ($4.25$) é menor que o valor crítico ($7.81$).

**Conclusão:** Não se rejeita $H_0$, confirmando que os dados seguem a distribuição exponencial proposta.